docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,105 @@
+# Freeing objects and resources
+
+* Status: Propsed
+* Deciders: ?
+* Date: 2023-10-10
+
+Prevous discussion: [PR 1394](https://github.com/mozilla/uniffi-rs/issues/1394), [Issue 8](https://github.com/mozilla/uniffi-rs/issues/8)
+
+## Context and Problem Statement
+
+UniFFI must support objects that hold resources that should be cleaned up promptly like database connections, file handles, network sockets, large memory allocations, etc.
+This is difficult, because objects like this are handled differently by different languages:
+
+* On Kotlin and other GC-based languages, the general advice is to explicitly close the resources and not rely on garbage collection to deal with them.
+* On Rust and other RAII-based languages, the general advice is to close the resources in a destructor.
+
+Currently, our Kotlin bindings try to address this problem for Objects by adding a `destroy()` method that manually drops the Rust value.
+They implement the `AutoCloseable` interface which allows users to ensure this is called via the `use` extension function.
+However, this is not sufficient for several reasons:
+
+  * It is specific to Kotlin, but many languages face similar issues.
+  * All objects get a resource closing functionality, when only some need it.
+  * This system is tied to decision to *not* support freeing UniFFI objects via GC-based systems like `Cleaner` (#8).
+    However, in some cases GC-based cleanup is what users want (#1394).
+  * It only solves the issue in one direction, however UniFFI also needs to deal with the other direction:
+    a Kotlin resource-holding object can be passed to a Rust library via a callback interface.
+    In this scenario, there's currently no mechanism for the Rust code to release the Kotlin object's resource.
+
+We should come up with a general approach for resource-holding objects and apply it consistently across all languages.
+
+## Decision Drivers
+
+* If an object does not hold a resource, then we should be able to use the GC to free it.
+* The solution should work for both Rust objects passed to the foreign language and foreign objects passed to Rust.
+
+## Considered Options
+
+### [Option 1] User-defined close method
+
+* Don't automatically implement any methods to close a resource.
+* Advise users to define a `close()` or similar method for resource-holding objects / callback
+  interfaces / trait interfaces and ensure it's called.  This also likely means defining an error
+  variant like `Error::Closed` that gets returned any methods that use the resource after the
+  `close()` method is called.  Alternatively, handle this by panicking or throwing an error.
+* Use the `Cleaner` API on Kotlin to automatically free UniFFI pointers for Kotlin objects.
+  Note that this combines with the previous bullet.  `close()` releases the resource, the Cleaner
+  action frees the heap allocation.
```

Comment:
I didn't know anything about this API, nor much about "oldschool" finalizers, but https://stackoverflow.com/questions/52879761/should-java-9-cleaner-be-preferred-to-finalization points towards `Cleaner` really being the best choice here.